source/docs/software/basic-programming/joystick.rst

```diff
@@ -0,0 +1,151 @@
+.. include:: <isonum.txt>
+
+Joysticks
+=========
+
+In FRC\ |reg|, almost any "controller" that can be recognized by Windows can be used as a joystick.  Joysticks are accessed using the ``GenericHID`` class.  This class has two relevant subclasses for preconfigured joysticks.  You may also implement your own for other controllers by extending ``GenericHID``.  The first is ``Joystick`` which is useful for standard flight joysticks.  The second is ``XboxController`` which works for the Xbox 360, Xbox One, or Logitech F310 (in XInput mode).  Each axis of the controller ranges from -1 to 1.
```

Comment:
I would prefer to start with ``Joystick`` and ``XboxController`` since those are what most people use, and then introduce the use of ``GenericHID``. Since it's an abstract class, it can't be used directly. Further, since ``Joystick`` gives access to all the raw values, it isn't necessary to use ``GenericHID`` even for non-joysticks.

Reply:
I don't understand what you are asking for?

I 100% disagree with using joystick for non flight controllers.  Sure you can do it but it is poor practice.

Reply:
Why is it poor practice?

Lets say you wanted to use a guitar hero controller and implement the red fret button and the whammy bar (axis)

1. You could use the Joystick or XboxController class and call getRawButton. getRawButtonPressed, getRawButtonReleased and getRawAxis with constants for which button is red and what axis is the whammy bar.
2. You could extend the Joystick or XboxController class and add methods boolean getRed(), boolean getRedPressed(), boolean getRedReleased() and double getWhammy()
3. You could extend the GenericHID class and add methods boolean getRed(), boolean getRedPressed(), boolean getRedReleased() and double getWhammy(). Since GenericHID is abstract, you also have to implement double getX​(GenericHID.Hand hand) and double getY​(GenericHID.Hand hand).

I think you could argue about whether 1 or 2 is better, but I don't see any way that 3 is better then 2, but 3 is what the article is suggesting.

Reply:
I do think 3 is the proper thing to do here.  

Maybe I am in the minority I don't know.  If so I think we should change the way GenericHID is set up.  Joystick has so much flight controller specific baggage I don't see how / why we would advocate for everyone extending from that just because they don't have to implement the abstracts.

Reply:
Implementing GenericHID seems like too much effort for a team to be expected to do just to use a controller.

Reply:
If GenericHID was concrete and didn't have the getX & Y methods, I would agree with you.

FYI, AP Java no longer covers abstract classes. https://apcentral.collegeboard.org/pdf/ap-computer-science-a-updates-2019-20.pdf

Reply:
Lets discuss Thursday.